docs/proposals/drive-replacement.md

```diff
@@ -0,0 +1,86 @@
+# Drive replacement procedure
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)  
+- [Design Details](#design-details)  
+<!-- /toc -->
+## Summary
+Disk replacement is a feature which allows user to physically replace defective drive.
+
+## Motivation
+Once specific disk goes offline CSI driver should provide a way to gracefully detach it from the system and insert new healthy device instead.
+
+Single drive can be used to store one or more persistent volumes. Persistent volumes can be owned by different applications which, in turn, can be managed by different Operators.
+
+### Goals
+- Develop algorithm for reactive and proactive drive replacement.
+- Come up with the API between CSI, Operator and User.
+
+## Proposal
+To perform replacement of failed drive negotiation between CSI, Operator and User is required:
+
+* BM CSI
+  - Detects drive health
+  - Triggers replacement process
+  - Sends alerts
+  - Prepares drive for physical replacement
+* Operator
+  - Subscribes for procedure support by setting corresponding PVC annotation.
+  - Instantiates recovery process and monitor its progress if needed
+  - Might reject procedure if system cannot release corresponding volume(s)
+  - Deletes persistent volumes
+* User
+  - Triggers physical drive replacement process since drive might not be shipped yet when recovery of data is completed
+## Design Details
+### Drive health detection
+Drive health is detected by drive manager and stored in the health field of [Drives CRD](https://github.com/dell/csi-baremetal/blob/master/charts/baremetal-csi-plugin/crds/baremetal-csi.dellemc.com_drives.yaml): 
+- `GOOD` - drive is healthy. Application can safely use it.
+- `SUSPECT` - drive might not be healthy. Replacement is recommended.
+- `BAD` - drive is not healthy. Replacement is required.
+- `UNKNOWN` - drive health not detected. Application should rely on IO errors.
+### Drive operational statuses
+[Drives CRD](https://github.com/dell/csi-baremetal/blob/master/charts/baremetal-csi-plugin/crds/baremetal-csi.dellemc.com_drives.yaml) will be extended by the new field - operational status: 
+- `OPERATIVE` - drive in use
+- `RELEASING` - releasing in progress
+- `RELEASED` - drive ready for removal
+- `FAILED` - drive failed to remove
+- `REMOVING` - removing in progress (PVs deletion, Safe remove, LED locate)
+- `REMOVED` - drive removed
+### API
+#### Operator
+Negotiation between CSI and Operator is optional and will be done though annotation on different objects.
+* Storage class annotation:
```

Comment:
Storage class it is a cluster level component which could be used by different operators. Does it mean that all who works with such SC must support volume release feature?

Reply:
This annotation must be set for PVC. I'll update this part.

Reply:
Updated.